src/access/Subjects.h

```diff
@@ -0,0 +1,208 @@
+    bool operator==(const PaseSubject & that) const { return true; }
+};
+
+struct NodeSubject
+{
+    NodeSubject(NodeId aNodeId) : nodeId(aNodeId) {}
+    NodeId nodeId;
+    bool operator==(const NodeSubject & that) const { return this->nodeId == that.nodeId; }
+};
+
+struct GroupSubject
+{
+    GroupSubject(GroupId aGroupId) : groupId(aGroupId) {}
+    GroupId groupId;
+    bool operator==(const GroupSubject & that) const { return this->groupId == that.groupId; }
+};
+
+class OperationalNodeId;
+
+/** A scoped subject is a unique identifier with-in a fabric scope. */
+class ScopedSubject
+{
+public:
+    ScopedSubject() {}
+
+    template <typename T, class... Args>
+    constexpr explicit ScopedSubject(InPlaceTemplateType<T>, Args &&... args) :
+        mSubject(InPlaceTemplate<T>, std::forward<Args>(args)...)
+    {}
+
+    template <typename T, typename... Args>
+    static ScopedSubject Create(Args &&... args)
+    {
+        return ScopedSubject(InPlaceTemplate<T>, std::forward<Args>(args)...);
+    }
+
+    Access::AuthMode GetAuthMode() const
+    {
+        if (mSubject.Is<PaseSubject>())
+        {
+            return Access::AuthMode::kPase;
+        }
+        else if (mSubject.Is<NodeSubject>())
+        {
+            return Access::AuthMode::kCase;
+        }
+        else if (mSubject.Is<GroupSubject>())
+        {
+            return Access::AuthMode::kGroup;
+        }
+        else
+        {
+            return Access::AuthMode::kNone;
+        }
+    }
+
+#if CHIP_DETAIL_LOGGING
+    const char * GetAuthModeString() const
+    {
+        switch (GetAuthMode())
+        {
+        case Access::AuthMode::kPase:
+            return "Pase";
+        case Access::AuthMode::kCase:
+            return "Case";
+        case Access::AuthMode::kGroup:
+            return "Group";
+        case Access::AuthMode::kNone:
+        default:
+            return "None";
+        }
+    }
+#endif // CHIP_DETAIL_LOGGING
+
+    /** Return subject value described in spec */
+    uint64_t GetValue() const
+    {
+        if (mSubject.Is<PaseSubject>())
+        {
+            return static_cast<uint64_t>(mSubject.Get<PaseSubject>().kPasscodeId) << 48;
+        }
+        else if (mSubject.Is<NodeSubject>())
+        {
+            return mSubject.Get<NodeSubject>().nodeId;
+        }
+        else if (mSubject.Is<GroupSubject>())
+        {
+            return static_cast<uint64_t>(mSubject.Get<GroupSubject>().groupId) << 48;
+        }
+        else
+        {
+            return std::numeric_limits<uint64_t>::max();
+        }
+    }
+
+    bool operator==(const ScopedSubject & that) const { return this->mSubject == that.mSubject; }
+
+private:
+    Variant<PaseSubject, NodeSubject, GroupSubject> mSubject;
+};
+
+/**
+ * A subject is a global unique identifier. It suite 2 purpose:
+ *
+ *  1. Identify an entity. operator== can be used to check if 2 subjects are identical.
+ *  2. Associate to ACL entries to grant privileges
```

Comment:
But it can't actually do that, because it does not carry the information that ACL needs (e.g. CATs, for CASE).

Reply:
`CATs` is another type of subject, I'll add it in the PR.

Reply:
This brings us back to the question of "what is this thing trying to represent and why?".  For ACL purposes a "subject" in the sense of an ACL entry can be one of a variety of things, but a "subject" in the sense of "what gets matched against an ACL entry" can be a list of things.

We don't have a singular concept of "subject" around.

Reply:
Like Boris, I'd like to understand what problem these new classes are trying to solve. Currently (and not judging whether it's ideal), a NodeId is a 64-bit value that, via spec, has its range partitioned into a few "kinds" of NodeId. We have some headers (NodeId.h and friends) with constants and functions for manipulating these NodeIds.

A subject is similar, but not exactly the same thing. ACL entries have subjects, and the SubjectDescriptor is provided by the messaging layer (those session objects and so on) through the IM to the Access Control system module, for performing checks during actions. Part of performing those checks is comparing ACL entry subjects against the SubjectDescriptor.

The other place subjects appear is in cluster operations (read/write entries), where they are just numbers, and the auth mode serves as a discriminant. There's no need for more type richness there, because we're not trying to protect naive users from writing a wrong subject, we expect admins to know what they are doing, and the values are validated regardless.

Looking at the linked issue, it seems the desire is to ensure all the info in those various sessions (in the messaging layer) is correct. This is good, but I think we only need to check all the places where a SubjectDescriptor is created, to ensure it is being fed appropriate messaging layer data.

The key thing to remember from a design perspective is this: subject is an access control construct, not a messaging construct. It is derived from messaging constructs, but is not the same thing. That's why it's defined in the access module, but created in the messaging module. It's an input API to the access control system module.